flink-core/src/main/java/org/apache/flink/api/java/typeutils/runtime/PojoComparatorGenerator.java

```diff
@@ -0,0 +1,207 @@
+		}
+		StringBuilder normalizableKeys = new StringBuilder();
+		for (int i = 0; i < comparators.length; ++i) {
+			normalizableKeys.append(String.format("if (f%d.supportsNormalizedKey()) {\n" +
+				"	if (f%d.invertNormalizedKey() != inverted) break;\n" +
+				"	nKeys++;\n" +
+				"	final int len = f%d.getNormalizeKeyLen();\n" +
+				"	this.normalizedKeyLengths[%d] = len;\n" +
+				"	nKeyLen += len;\n" +
+				"	if (nKeyLen < 0) {\n" +
+				"		nKeyLen = Integer.MAX_VALUE;\n" +
+				"		break;\n" +
+				"	}\n" +
+				"} else {\n" +
+				"	break;\n" +
+				"}\n", i, i, i, i));
+		}
+		StringBuilder cloneMembers = new StringBuilder();
+		for (int i = 0; i < comparators.length; ++i) {
+			cloneMembers.append(String.format("f%d = toClone.f%d.duplicate();\n", i, i));
+		}
+		StringBuilder flatComparators = new StringBuilder();
+		for (int i = 0; i < comparators.length; ++i) {
+			flatComparators.append(String.format(
+				"if(f%d instanceof CompositeTypeComparator) {\n" +
+				"	((CompositeTypeComparator)f%d).getFlatComparator(flatComparators);\n" +
+				"} else {\n" +
+				"	flatComparators.add(f%d);\n" +
+				"}\n", i, i, i));
+		}
+		StringBuilder hashMembers = new StringBuilder();
+		for (int i = 0; i < keyFields.length; ++i) {
+			hashMembers.append(String.format(
+				"code *= TupleComparatorBase.HASH_SALT[%d & 0x1F];\n" +
+				"code += this.f%d.hash(((" + typeName + ")value)." + accessStringForField(keyFields[i]) +
+					");\n",
+				i, i));
+		}
+		StringBuilder setReference = new StringBuilder();
+		for (int i = 0; i < comparators.length; ++i) {
+			setReference.append(String.format(
+				"this.f%d.setReference(((" + typeName + ")toCompare)." + accessStringForField(keyFields[i]) + ");\n",
+				i));
+		}
+		StringBuilder equalToReference = new StringBuilder();
+		for (int i = 0; i < comparators.length; ++i) {
+			equalToReference.append(String.format(
+				"if (!this.f%d.equalToReference(((" + typeName + ")candidate)." +
+				accessStringForField(keyFields[i]) + ")) {\n" +
+				"	return false;\n" +
+				"}\n", i));
+		}
+		StringBuilder compareToReference = new StringBuilder();
+		for (int i = 0; i < comparators.length; ++i) {
+			compareToReference.append(String.format(
+				"cmp = this.f%d.compareToReference(other.f%d);\n" +
+				"if (cmp != 0) {\n" +
+				"	return cmp;\n" +
+				"}\n", i, i));
+		}
+		StringBuilder compareFields = new StringBuilder();
+		for (int i = 0; i < comparators.length; ++i) {
+			compareFields.append(String.format(
+				"cmp = f%d.compare(((" + typeName + ")first)." + accessStringForField(keyFields[i]) + "," +
+				"((" + typeName + ")second)." + accessStringForField(keyFields[i]) + ");\n" +
+				"if (cmp != 0) {\n" +
+					"return cmp;\n" +
+				"}\n", i));
+		}
+		StringBuilder putNormalizedKeys = new StringBuilder();
+		for (int i = 0; i < comparators.length; ++i) {
+			putNormalizedKeys.append(String.format("if (%d >= numLeadingNormalizableKeys || numBytes <= 0) break;\n" +
```

Comment:
Instead of the  `%d >= numLeadingNormalizableKeys` check, you could have the for loop go up to `numLeadingNormalizableKeys`.

Also, why is this a `break` (while false loop in the ftl instead of just returning here?)